.github/workflows/deploy.yaml

```diff
@@ -9,45 +9,189 @@ on:
+        with:
+          python-version: "3.x"
+
+      - name: Download dist files
+        uses: actions/download-artifact@v4
+        with:
+          name: dist-files
+          path: dist/
+
+      - name: Install package with test dependencies
+        run: |
+          python -m pip install --upgrade pip
+          WHEEL_FILE=$(ls dist/*.whl)
+          pip install "$WHEEL_FILE"[test]
+
+      - name: Run tests
+        run: |
+          pytest -n auto --verbose --pyargs basicrta
+
+  deploy-testpypi:
+    name: Deploy to TestPyPI
+    runs-on: ubuntu-latest
+    needs: [build, test-install, test-pytest]
+    if: |
+      github.repository == 'Becksteinlab/basicrta' &&
+      (github.event_name == 'push' && startsWith(github.ref, 'refs/tags/'))
     environment:
-      name: deploy
-      url: https://pypi.org/p/basicrta
+      name: testpypi
+      url: https://test.pypi.org/p/basicrta
     permissions:
-      id-token: write
+      id-token: write  # IMPORTANT: mandatory for trusted publishing
+    steps:
+      - name: Download dist files
+        uses: actions/download-artifact@v4
+        with:
+          name: dist-files
+          path: dist/
+
+      - name: Publish to TestPyPI
+        uses: pypa/gh-action-pypi-publish@v1.12.4
+        with:
+          repository-url: https://test.pypi.org/legacy/
+
+  deploy-pypi:
+    name: Deploy to PyPI
+    runs-on: ubuntu-latest
+    needs: [build, test-install, test-pytest]
     if: |
       github.repository == 'Becksteinlab/basicrta' &&
       (github.event_name == 'release' && github.event.action == 'published')
-    name: Build, upload and test pure Python wheels to PyPi
-    runs-on: ubuntu-latest
+    environment:
+      name: pypi
+      url: https://pypi.org/p/basicrta
+    permissions:
+      id-token: write  # IMPORTANT: mandatory for trusted publishing
+    steps:
+      - name: Download dist files
+        uses: actions/download-artifact@v4
+        with:
+          name: dist-files
+          path: dist/
+
+      - name: Publish to PyPI
+        uses: pypa/gh-action-pypi-publish@v1.12.4
 
+  test-deployed-testpypi:
+    name: Test deployed package (TestPyPI)
+    runs-on: ubuntu-latest
+    needs: deploy-testpypi
+    if: |
+      github.repository == 'Becksteinlab/basicrta' &&
+      (github.event_name == 'push' && startsWith(github.ref, 'refs/tags/'))
     steps:
-      - uses: actions/checkout@v4
+      - name: Set up Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.x"
+
+      - name: Wait for package to be available from testpypi
+        run: sleep 10
 
-      - name: pypi_deploy
-        uses: MDAnalysis/pypi-deployment@main
-        if: github.event_name == 'release' && github.event.action == 'published'
+      - name: Install from TestPyPI
+        run: |
+          python -m pip install --upgrade pip
+          pip install --index-url https://test.pypi.org/simple/ --extra-index-url https://pypi.org/simple/ basicrta[test]
```

Comment:
This works but does not install pre-release versions. Keep in mind when testing.